### PR TITLE
Fixes for UITextView rx_text bug with autocorrect enabled #333

### PR DESCRIPTION
**Pull request by sergdort**

### CODE REVIEW

**Comment by kzaher** [comment]
Hi @sergdort ,

I think we can probably pull in your solution after RC, before 2.0.0 release, assuming it works :) (since current solution works, and this solution doesn't remove anything from public API)

I'll still need to try it for myself.

I have noted a couple of issues we would want to tackle before merging this in develop branch, I'll note my comments in the code (and build server error in unit tests is one of them :)

File: RxCocoa/iOS/NSTextStorage+Rx.swift [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    import RxSwift
+#endif
+
+extension NSTextStorage {
+    
+    public var rx_delegate:DelegateProxy {
+        return proxyForObject(RxTextStorageDelegateProxy.self, self)
+    }
+    
+    public var rx_string:Observable<String> {
```
Comment (kzaher) [inline RxCocoa/iOS/NSTextStorage+Rx.swift]: The convention is for delegate proxies to expose a method that is named the same like corresponding delegate method.

You can find samples in other delegate proxies.

File: RxCocoa/iOS/NSTextStorage+Rx.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        return rx_delegate
+            .observe("textStorage:didProcessEditing:range:changeInLength:")
+            .map({ a  in
+                let textStorage:NSTextStorage = castOrFatalError(a[0])
+                return textStorage.string
+            })
+            .distinctUntilChanged() // dont know why, but system call delegate twice on auto correction
```
Comment (kzaher) [inline RxCocoa/iOS/NSTextStorage+Rx.swift]: Last sentence was wrong :) We don't expose non required methods in delegate proxies :)

The `rx_text` method inside `UITextView+Rx.swift` should look something like 

```swift
let source: Observable<String> = Observable.deferred { [weak self] () -> Observable<String> in
            let text = self?.text ?? ""
            // basic event
            let textChangedEvent = self?.textStorage..observe("textStorage:didProcessEditing:range:changeInLength:") ?? Observable.empty()

 return textChangedEvent
                .map { a in
                    return self?.text ?? ""
                }
                .startWith(text)
                .distinctUntilChanged()
            }
}
```

File: RxCocoa/iOS/NSTextStorage+Rx.swift (hunk 1 above)
Comment (sergdort) [inline RxCocoa/iOS/NSTextStorage+Rx.swift]: Do you mean `rx_string` ?

Comment (kzaher) [inline RxCocoa/iOS/NSTextStorage+Rx.swift]: This is snippet from CLLocationManager
```swift
/**
    Reactive wrapper for `delegate` message.
    */
    public var rx_rangingBeaconsDidFailForRegionWithError: Observable<(region: CLBeaconRegion, error: NSError)> {
        return rx_delegate.observe("locationManager:rangingBeaconsDidFailForRegion:withError:")
            .map { a in
                let region = try castOrThrow(CLBeaconRegion.self, a[1])
                let error = try castOrThrow(NSError.self, a[2])
                return (region: region, error: error)
            }
    }
```

so in this case this should probably be something like
```swift
public var rx_didProcessEditingRangeChangeInLength: Observable<(editedMask:NSTextStorageEditActions, editedRange:NSRange, delta:)> {
       ....
}
```

File: RxCocoa/iOS/NSTextStorage+Rx.swift (hunk 2 above)
Comment (sergdort) [inline RxCocoa/iOS/NSTextStorage+Rx.swift]: I see, thanks. Will do

File: RxCocoa/iOS/NSTextStorage+Rx.swift (hunk 1 above)
Comment (sergdort) [inline RxCocoa/iOS/NSTextStorage+Rx.swift]: Thanks :)

**Comment by kzaher** [comment]
Great we've found a solution that isn't hacky.

**Pull request merged**